src/scalar_mul/pippenger.rs

```diff
@@ -0,0 +1,211 @@
+        } else if size < 23_000 {
+            11
+        } else if size < 45_000 {
+            12
+        } else if size < 160_000 {
+            13
+        } else if size < 260_000 {
+            14
+        } else {
+            15
+        };
+
+        let window_size = 1 << w;
+        let windows_count = (256 + w - 1) / w; // == ceil(256/w)
+        let buckets_count = window_size / 2; // digits are signed+centered hence 2^w/2, excluding 0-th bucket
+
+        // Step 1. Prepare scalar representation using signed digits radix 2^w.
+        // Each digit will be in [-2^w/2, 2^w/2].
+        let all_digits: Vec<_> = scalars
+            .into_iter()
+            .map(|c| c.borrow().to_arbitrary_radix(w))
+            .collect();
+        let ps: Vec<ProjectiveNielsPoint> = points
+            .into_iter()
+            .map(|p| p.borrow().to_projective_niels())
+            .collect();
+
+        // Iterate the digits from last to first so we can shift the result by w bits
+        // at the end of each iteration (e.g. w doublings).
+        let mut result = EdwardsPoint::identity();
+        for i in (0..windows_count).rev() {
+            // Step 2. Shift the current result by w bits to leave room for a new window.
+            if i < (windows_count - 1) {
+                result = result.mul_by_pow_2(w as u32);
+            }
+
+            // Step 3. Prepare 2^w/2 buckets.
+            // buckets[i] corresponds to a multiplication factor (i+1).
+            let mut buckets: Vec<_> = (0..buckets_count)
+                .map(|_| EdwardsPoint::identity())
+                .collect();
+
+            // Step 4. Iterate over pairs of (point, scalar)
+            // and add/sub the point to the corresponding bucket.
+            // Note: when we add support for precomputed lookup tables,
+            // we'll be adding/subtractiong point premultiplied by `digits[i]` to buckets[0].
+            for (digits, pt) in all_digits.iter().zip(ps.iter()) {
+                if digits[i] > 0 {
+                    let b = (digits[i] - 1) as usize;
+                    buckets[b] = (&buckets[b] + pt).to_extended();
+                } else if digits[i] < 0 {
+                    let b = (-digits[i] - 1) as usize;
+                    buckets[b] = (&buckets[b] - pt).to_extended();;
+                }
+            }
+
+            // Step 5. Add the buckets applying the multiplication factor to each bucket.
+            // The most efficient way to do that is to have a single sum with two running sums:
+            // an intermediate sum from last bucket to the first, and a sum of intermediate sums.
+            //
+            // For example, to add buckets 1*A, 2*B, 3*C we need to add these points:
+            //   C
+            //   C B
+            //   C B A   Sum = C + (C+B) + (C+B+A)
+            let mut buckets_intermediate_sum = buckets[buckets_count - 1].clone();
+            let mut buckets_sum = buckets[buckets_count - 1].clone();
```

Comment:
It would be good to do allocate the scratch space for this up-front, rather than doing two allocations per loop iteration

Reply:
Good point, fixed in 67a6a97.